File: sort/quick_sort/julia/quick_sort.jl

function swap(array, i, j)
  tmp = array[i]
  array[i] = array[j]
  array[j] = tmp
end

function partition(array, left, right)
  pivotIndex = rand(left:right)
  pivot = array[pivotIndex]
  swap(array, pivotIndex, right)

  pos = left
  for i = left:right-1
    if array[i] < pivot
      swap(array, i, pos)
      pos = pos+1
    end
  end
  
  swap(array, pos, right)

  return pos
end

function quick_sort(array, left, right)
  if left < right
    pivotIndex = partition(array, left, right)
    quick_sort(array, left, pivotIndex-1)
    quick_sort(array, pivotIndex+1, right)
  end
end


function is_sorted(array)
  max = 0
  i = 1
  while i < length(array)
    if array[i] < max
      return false
    end
    max = array[i]
    i = i+1
  end
  return true
end

function generate_random_array(length)
  rand(0:length, length)
end


for i = 1:100
  array = generate_random_array(1000)
  quick_sort(array, 1, length(array))
  if !is_sorted(array)
    println("Error! Array is not sorted!")
    exit()
  end
end

println("Quick sort works fine!")

File: sort/quick_sort/go/quicksort.go
package quicksort

// QuickSort uses the Quick sort algorithm to sort an integer slice.
// The following code implements Lomuto partition scheme, based on
// wall shifting idea and median-of-three pivot selection algorithm
// for higher optimality. You can read more about that here:
// https://en.wikipedia.org/wiki/Quicksort#Lomuto_partition_scheme
// https://en.wikipedia.org/wiki/Quicksort#Choice_of_pivot
func QuickSort(slice []int) {
	quickSort(slice, 0, len(slice)-1)
}

func quickSort(slice []int, from, to int) {
	if from < to {
		mid := partition(slice, from, to)
		quickSort(slice, mid+1, to)
		quickSort(slice, from, mid-1)
	}
}

func getMedian(slice []int, left, right int) int {
	mid := left + (right-left)/2

	if slice[left] < slice[mid] {
		if slice[right] < slice[left] {
			return left
		} else if slice[right] < slice[mid] {
			return right
		}
		return mid
	}
	if slice[right] < slice[mid] {
		return mid
	} else if slice[right] < slice[left] {
		return right
	}
	return left
}

func partition(slice []int, from, to int) int {
	pivot := getMedian(slice, from, to)
	wall := to

	slice[pivot], slice[from] = slice[from], slice[pivot]

	for nowPos := to; nowPos > from; nowPos-- {
		if slice[from] < slice[nowPos] {
			slice[nowPos], slice[wall] = slice[wall], slice[nowPos]
			wall--
		}
	}
	slice[from], slice[wall] = slice[wall], slice[from]
	return wall
}
